node/src/components/consensus/highway_core/state.rs

```diff
@@ -361,54 +363,47 @@ impl<C: Context> State<C> {
         }
-        match wvote.panorama.get(creator) {
-            Observation::Faulty => {
-                error!("Vote from faulty validator should be rejected in `pre_validate_vote`.");
-                return Err(VoteError::FaultyCreator); // Should be unreachable.
+        if wvote.seq_number != panorama.next_seq_num(self, creator) {
+            return Err(VoteError::SequenceNumber);
+        }
+        let opt_prev_vote = panorama[creator].correct().map(|vh| self.vote(vh));
+        let prev_round_exp = opt_prev_vote.map_or(self.params.init_round_exp(), |v| v.round_exp);
+        let max_re = prev_round_exp.max(wvote.round_exp);
+        // The round exponent must only change one step at a time.
+        if prev_round_exp + 1 < max_re || wvote.round_exp + 1 < max_re {
+            return Err(VoteError::RoundLength);
+        }
+        if let Some(prev_vote) = opt_prev_vote {
+            if prev_round_exp != wvote.round_exp
+                && prev_vote.timestamp >> max_re == timestamp >> max_re
+            {
+                return Err(VoteError::RoundLength);
             }
-            Observation::None if wvote.seq_number == 0 => (),
-            Observation::None => return Err(VoteError::SequenceNumber),
-            Observation::Correct(hash) => {
-                let prev_vote = self.vote(hash);
-                // The sequence number must be one more than the previous vote's.
-                if wvote.seq_number != 1 + prev_vote.seq_number {
-                    return Err(VoteError::SequenceNumber);
-                }
-                // The round exponent must only change one step at a time, and not within a round.
-                if prev_vote.round_exp != wvote.round_exp {
-                    let max_re = prev_vote.round_exp.max(wvote.round_exp);
-                    if prev_vote.round_exp + 1 < max_re
-                        || wvote.round_exp + 1 < max_re
-                        || prev_vote.timestamp >> max_re == wvote.timestamp >> max_re
-                    {
-                        return Err(VoteError::RoundLength);
-                    }
-                }
-                let r_id = round_id(wvote.timestamp, wvote.round_exp);
-                if let Some(prev2_vote) = prev_vote.previous().map(|h2| self.vote(h2)) {
-                    if prev2_vote.round_id() == r_id {
-                        return Err(VoteError::ThreeVotesInRound);
-                    }
-                }
-                if wvote.value.is_some()
-                    && (prev_vote.round_id() == r_id
-                        || wvote.timestamp != r_id
-                        || self.leader(r_id) != creator)
-                {
-                    return Err(VoteError::NonLeaderBlock);
+            let r_id = round_id(timestamp, wvote.round_exp);
+            if wvote.value.is_some()
+                && (prev_vote.round_id() == r_id
+                    || timestamp != r_id
+                    || self.leader(r_id) != creator)
```

Comment:
Do we need `self.leader(r_id) != creator` at all? Sending a second block in the same round is invalid even for the round's leader.

Reply:
But this is a disjunction. If:
* the previous vote was in the same round, _or_
* the vote's timestamp is not the beginning of the round, _or_
* we are not the leader,

then this vote must not be a block.

Reply:
Ah, right…

Reply:
But actually, these checks should happen even if this was the first vote! Fixing that…

Reply:
And that actually broke the test, because the first leader is, in fact, Alice, not Bob!

Reply:
nice find!